# Task 3:

Train a convolutional neural network on the SVHN dataset (http://ufldl.stanford.edu/housenumbers/) in format 2 (single digit classification). You should achieve at least 85% test-set accuracy with a base model. Also build a model using batch normalization. 

### The Street View House Numbers (SVHN) Dataset

SVHN is obtained from house numbers in Google Street View images. SVHN is a real-world image dataset for developing machine learning and object recognition algorithms with minimal requirement on data preprocessing and formatting. 10 classes, 1 for each digit. Digit '1' has label 1, '9' has label 9 and '0' has label 10.
73257 digits for training, 26032 digits for testing, and 531131 additional, somewhat less difficult samples, to use as extra training data
Comes in two formats:
1. Original images with character level bounding boxes.
2. MNIST-like 32-by-32 images centered around a single character (many of the images do contain some distractors at the sides).



In [0]:
# Importing libraries

import numpy as np
import pandas as pd
import tensorflow
from keras.layers import BatchNormalization
from keras.wrappers.scikit_learn import KerasClassifier, KerasRegressor
from keras.layers import Conv2D, MaxPooling2D, Flatten
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
import keras
import tempfile
import scipy
import scipy.io as sio
import os
 

In [0]:
# Importing dataset

train = scipy.io.loadmat('train_32x32.mat')
test = scipy.io.loadmat('test_32x32.mat')

In [0]:
train

{'__header__': b'MATLAB 5.0 MAT-file, Platform: GLNXA64, Created on: Mon Dec  5 21:09:26 2011',
 '__version__': '1.0',
 '__globals__': [],
 'X': array([[[[ 33,  84,  19, ...,  92, 190, 216],
          [ 30,  76,  54, ...,  78, 188, 217],
          [ 38,  59, 110, ..., 101, 191, 212]],
 
         [[ 15,  86,  20, ...,  94, 205, 221],
          [ 23,  73,  52, ...,  82, 203, 222],
          [ 19,  66, 111, ..., 105, 206, 217]],
 
         [[ 15,  77,  25, ..., 114, 220, 226],
          [ 17,  78,  57, ..., 101, 218, 227],
          [ 19,  56, 116, ..., 125, 220, 221]],
 
         ...,
 
         [[ 72,  90,  65, ..., 200, 229, 200],
          [ 65,  78, 144, ..., 201, 231, 199],
          [ 56,  69, 223, ..., 203, 224, 191]],
 
         [[ 82,  88,  78, ..., 192, 229, 193],
          [ 77,  77, 148, ..., 193, 229, 188],
          [ 57,  67, 218, ..., 195, 224, 182]],
 
         [[ 89,  88,  98, ..., 190, 229, 197],
          [ 79,  78, 158, ..., 191, 228, 189],
          [ 59,  66, 220, 

In [0]:
test

{'__header__': b'MATLAB 5.0 MAT-file, Platform: GLNXA64, Created on: Mon Dec  5 21:18:15 2011',
 '__version__': '1.0',
 '__globals__': [],
 'X': array([[[[ 38, 129, 150, ..., 115,  96, 101],
          [103, 142, 160, ..., 132,  65,  75],
          [ 60, 153, 169, ..., 142,  47,  60]],
 
         [[ 39, 127, 150, ..., 116,  97, 100],
          [104, 143, 163, ..., 133,  65,  73],
          [ 61, 152, 170, ..., 143,  49,  60]],
 
         [[ 39, 125, 152, ..., 117,  97,  99],
          [104, 143, 168, ..., 134,  65,  71],
          [ 62, 151, 172, ..., 144,  50,  59]],
 
         ...,
 
         [[ 41, 121, 153, ..., 114,  86,  95],
          [102, 133, 172, ..., 136,  61,  63],
          [ 61, 153, 180, ..., 141,  54,  52]],
 
         [[ 42, 123, 150, ..., 114,  87,  95],
          [103, 134, 171, ..., 136,  62,  63],
          [ 62, 156, 181, ..., 142,  55,  52]],
 
         [[ 39, 123, 147, ..., 115,  88,  96],
          [ 97, 135, 169, ..., 136,  63,  64],
          [ 57, 157, 180, 

In [0]:
x_train = train['X']

In [0]:
y_train = train['y']

In [0]:
x_test = test['X']

In [0]:
y_test = test['y']

In [0]:
batch_size = 128
num_classes = 11
epochs = 12

In [0]:
# Input image dimensions
img_rows, img_cols = 32, 32

In [0]:
# Data, shuffled and split between train and test sets

x_train=np.rollaxis(x_train, 3, 0)
x_test=np.rollaxis(x_test,3,0)


In [0]:
# Defining the input shape
input_shape = (img_rows, img_cols, 3)

In [0]:
from keras.utils.np_utils import to_categorical

y_train_binary = to_categorical(y_train)
y_test_xbinary=to_categorical(y_test)


In [0]:
# Implementing the Base model

num_classes = 11
cnn = Sequential()

In [0]:
cnn.add(Conv2D(16, kernel_size=(5, 5),
                 activation='relu',
                 input_shape=input_shape))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Conv2D(32, (5, 5),activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Conv2D(64, (5, 5),activation='relu'))

In [0]:
cnn.add(Flatten())
cnn.add(Dropout(0.5))
cnn.add(Dense(16, activation='relu'))
cnn.add(Dense(num_classes, activation='softmax'))

In [0]:
cnn.compile("adam", "categorical_crossentropy", metrics=['accuracy'])

In [0]:
# Defining history by fitting cnn

history_cnn = cnn.fit(x_train, y_train_binary,
                      batch_size = 200, epochs=50, verbose=1, validation_split=.2)

Train on 58605 samples, validate on 14652 samples
Epoch 1/50
58605/58605 [==============================] - 122s 2ms/step - loss: 2.0201 - acc: 0.3310 - val_loss: 1.1757 - val_acc: 0.6490
Epoch 2/50
58605/58605 [==============================] - 110s 2ms/step - loss: 1.1658 - acc: 0.6210 - val_loss: 0.8290 - val_acc: 0.7519
Epoch 3/50
58605/58605 [==============================] - 108s 2ms/step - loss: 0.9437 - acc: 0.6994 - val_loss: 0.6419 - val_acc: 0.8168
Epoch 4/50
58605/58605 [==============================] - 111s 2ms/step - loss: 0.8494 - acc: 0.7327 - val_loss: 0.5768 - val_acc: 0.8356
Epoch 5/50
58605/58605 [==============================] - 114s 2ms/step - loss: 0.7746 - acc: 0.7557 - val_loss: 0.5796 - val_acc: 0.8346
Epoch 6/50
58605/58605 [==============================] - 111s 2ms/step - loss: 0.7375 - acc: 0.7671 - val_loss: 0.5328 - val_acc: 0.8490
Epoch 7/50
58605/58605 [==============================] - 108s 2ms/step - loss: 0.6988 - acc: 0.7793 - val_loss: 0.5213 - 

In [0]:
# Getting score for the base model

score_basic = cnn.evaluate(x_test, y_test_xbinary)


26032/26032 [==============================] - 18s 702us/step


In [0]:
# Printing the base model test accuracy

print("Task3 base model Test Accuracy: {:.2f}".format(score_basic[1]))

Task3 base model Test Accuracy: 0.87


In [0]:
# Batch Normalization

# Tiny convent with batch normalization

num_classes = 11
cnn_small_bn = Sequential()
cnn_small_bn.add(Conv2D(8, kernel_size=(3, 3),
                 input_shape=input_shape))
cnn_small_bn.add(Activation("relu"))
cnn_small_bn.add(BatchNormalization())
cnn_small_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_small_bn.add(Conv2D(8, (3, 3)))
cnn_small_bn.add(Activation("relu"))
cnn_small_bn.add(BatchNormalization())
cnn_small_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_small_bn.add(Flatten())
cnn_small_bn.add(Dense(64, activation='relu'))
cnn_small_bn.add(Dense(num_classes, activation='softmax'))

In [0]:
cnn_small_bn.compile("adam", "categorical_crossentropy", metrics=['accuracy'])

In [0]:
history_cnn_small_bn = cnn_small_bn.fit(x_train, y_train_binary,
                                        batch_size=200, epochs=50, verbose=1, validation_split=.2)

Train on 58605 samples, validate on 14652 samples
Epoch 1/50
58605/58605 [==============================] - 103s 2ms/step - loss: 1.5218 - acc: 0.5090 - val_loss: 0.8433 - val_acc: 0.7388
Epoch 2/50
58605/58605 [==============================] - 102s 2ms/step - loss: 0.6431 - acc: 0.8109 - val_loss: 0.6373 - val_acc: 0.8134
Epoch 3/50
58605/58605 [==============================] - 102s 2ms/step - loss: 0.5251 - acc: 0.8474 - val_loss: 0.7690 - val_acc: 0.7632
Epoch 4/50
58605/58605 [==============================] - 102s 2ms/step - loss: 0.4799 - acc: 0.8609 - val_loss: 0.5007 - val_acc: 0.8522
Epoch 5/50
58605/58605 [==============================] - 104s 2ms/step - loss: 0.4293 - acc: 0.8745 - val_loss: 0.5059 - val_acc: 0.8521
Epoch 6/50
58605/58605 [==============================] - 107s 2ms/step - loss: 0.4042 - acc: 0.8813 - val_loss: 0.4756 - val_acc: 0.8607
Epoch 7/50
58605/58605 [==============================] - 107s 2ms/step - loss: 0.3788 - acc: 0.8887 - val_loss: 0.5297 - 

In [0]:
# Evaluating Score with batch normalization

score_batch = cnn_small_bn.evaluate(x_test,y_test_xbinary)

26032/26032 [==============================] - 14s 531us/step


In [0]:
# Printing score means test accuracy of the batch model after batch normalization

print("Task3 batch model Test Accuracy: {:.2f}".format(score_batch[1]))

Task3 batch model Test Accuracy: 0.85


After observing the behaviour of the models as we can see from the above results, the test accuracy of without batch normalization for the base model is 87% which is pretty awesome for the convolutional neural network for SVHN dataset and the test accuracy for the batch model after batch normalization is 85 %.